File: matlab/tests/test_ma_rel_gate.m
%% test_ma_rel_gate.m
%  Regression tests for the centres-vs-grid gate on the multi-attribute
%  relative path (MOBIUS.MARELATTRPREFERSCENTRES). Mirrors the Python
%  tests/test_ma_rel_gate.py coverage cell-for-cell.
%
%  The gate compares predicted wall time on the two paths. Its
%  calibration notes (see the constants at the top of the .m file)
%  explain the empirical fit. The cells below pin its decisions on the
%  labelled truth set from that fit, including the r = 2 band where the
%  previous raw-op-count gate over-selected centres by one to two orders
%  of magnitude in wall time.
%
%  Each case is one (K, r_a, sigma, span_or_period, isPer, expectCentres)
%  tuple. The MATLAB and Python cost-model constants match exactly so
%  that the two languages route the same cells to the same path.
%
%  Standalone-runnable; appends to `results` when called from test_mpt.m.

if ~exist('results', 'var')
    results = {};
    standalone = true;
else
    standalone = false;
end

rng(0);

% --------------------------------------------------------------------
% Non-periodic r = 2 band -- previously the gate over-selected centres
% here; the corrected gate must return grid.
% --------------------------------------------------------------------
nonperR2Grid = { ...
    % K, sigma, spanPerSide
    20, 15.0, 1800.0; ...
    30, 15.0, 1800.0; ...
    50, 15.0, 1800.0; ...
    70, 15.0, 1800.0; ...
    30, 30.0, 1800.0; ...
    40,  5.0, 1800.0; ...
    50,  5.0, 1800.0; ...
};
for ii = 1:size(nonperR2Grid, 1)
    K = nonperR2Grid{ii, 1};
    sigma = nonperR2Grid{ii, 2};
    span = nonperR2Grid{ii, 3};
    Px = sort(rand(K, 1) * span);
    Py = sort(rand(K, 1) * span);
    tf = mobius.maRelAttrPrefersCentres(Px, Py, sigma, 2, true, false, 0);
    name = sprintf('gate non-per r=2: K=%d sigma=%.0f -> grid', K, sigma);
    results(end+1, :) = {name, ~tf}; %#ok<*SAGROW>
end

% --------------------------------------------------------------------
% Non-periodic cells where centres is genuinely faster.
% --------------------------------------------------------------------
nonperCentres = { ...
    % K, r, sigma, spanPerSide
     5, 2, 15.0, 1800.0; ...
     8, 2, 15.0, 1800.0; ...
    12, 2, 15.0, 1800.0; ...
     6, 3, 15.0, 1800.0; ...
     8, 3, 15.0, 1800.0; ...
     5, 4, 15.0, 1800.0; ...
     6, 4, 15.0, 1800.0; ...
};
for ii = 1:size(nonperCentres, 1)
    K = nonperCentres{ii, 1};
    r_a = nonperCentres{ii, 2};
    sigma = nonperCentres{ii, 3};
    span = nonperCentres{ii, 4};
    Px = sort(rand(K, 1) * span);
    Py = sort(rand(K, 1) * span);
    tf = mobius.maRelAttrPrefersCentres(Px, Py, sigma, r_a, true, false, 0);
    name = sprintf('gate non-per: K=%d r=%d -> centres', K, r_a);
    results(end+1, :) = {name, tf};
end

% --------------------------------------------------------------------
% Periodic cells where the gate must pick grid.
% --------------------------------------------------------------------
perGrid = { ...
    % K, r, sigma, period
    30, 2, 15.0, 3600.0; ...
    50, 2, 15.0, 3600.0; ...
    12, 3, 15.0, 3600.0; ...
    10, 3,  5.0, 1200.0; ...
     8, 4, 15.0, 3600.0; ...
};
for ii = 1:size(perGrid, 1)
    K = perGrid{ii, 1};
    r_a = perGrid{ii, 2};
    sigma = perGrid{ii, 3};
    P = perGrid{ii, 4};
    Px = sort(rand(K, 1) * P);
    Py = sort(rand(K, 1) * P);
    tf = mobius.maRelAttrPrefersCentres(Px, Py, sigma, r_a, true, true, P);
    name = sprintf('gate per: K=%d r=%d sigma=%.0f -> grid', K, r_a, sigma);
    results(end+1, :) = {name, ~tf};
end

% --------------------------------------------------------------------
% Periodic cells where centres is cheaper.
% --------------------------------------------------------------------
perCentres = { ...
    % K, r, sigma, period
     5, 2, 15.0, 3600.0; ...
     8, 2, 15.0, 3600.0; ...
     6, 3, 15.0, 3600.0; ...
     8, 3, 15.0, 3600.0; ...
     5, 4, 15.0, 3600.0; ...
     6, 4, 15.0, 3600.0; ...
};
for ii = 1:size(perCentres, 1)
    K = perCentres{ii, 1};
    r_a = perCentres{ii, 2};
    sigma = perCentres{ii, 3};
    P = perCentres{ii, 4};
    Px = sort(rand(K, 1) * P);
    Py = sort(rand(K, 1) * P);
    tf = mobius.maRelAttrPrefersCentres(Px, Py, sigma, r_a, true, true, P);
    name = sprintf('gate per: K=%d r=%d -> centres', K, r_a);
    results(end+1, :) = {name, tf};
end

% --------------------------------------------------------------------
% Trivial early-exit cases (invariant under any calibration).
% --------------------------------------------------------------------
PxAny = sort(rand(20, 1) * 1800);
PyAny = sort(rand(20, 1) * 1800);
results(end+1, :) = {'gate: absolute attribute -> false', ...
    ~mobius.maRelAttrPrefersCentres(PxAny, PyAny, 15, 2, false, false, 0)};
results(end+1, :) = {'gate: r_a < 2 -> false', ...
    ~mobius.maRelAttrPrefersCentres(PxAny, PyAny, 15, 1, true, false, 0)};

Px2 = sort(rand(2, 1) * 1800);
Py2 = sort(rand(2, 1) * 1800);
results(end+1, :) = {'gate: K < r_a (K=2,r=3) -> false', ...
    ~mobius.maRelAttrPrefersCentres(Px2, Py2, 15, 3, true, false, 0)};

Px3 = sort(rand(3, 1) * 1800);
Py3 = sort(rand(3, 1) * 1800);
results(end+1, :) = {'gate: K < r_a (K=3,r=4) -> false', ...
    ~mobius.maRelAttrPrefersCentres(Px3, Py3, 15, 4, true, false, 0)};

Px8 = sort(rand(8, 1) * 100);
Py8 = sort(rand(8, 1) * 100);
results(end+1, :) = {'gate: above sigma/P threshold -> false', ...
    ~mobius.maRelAttrPrefersCentres(Px8, Py8, 10, 2, true, true, 100)};


if standalone
    nPass = sum(cellfun(@(x) isequal(x, true), results(:, 2)));
    nFail = size(results, 1) - nPass;
    fprintf('test_ma_rel_gate: %d passed, %d failed\n', nPass, nFail);
    if nFail > 0
        for ii = 1:size(results, 1)
            if ~isequal(results{ii, 2}, true)
                fprintf('  FAIL  %s\n', results{ii, 1});
            end
        end
    end
end

File: python/tests/test_ma_rel_gate.py
"""Regression tests for the centres-vs-grid gate on the multi-attribute
relative path (``_ma_rel_attr_prefers_centres``).

The gate compares predicted wall time on the two paths. Its calibration
notes (see the constants ``_CENTRES_NS_BASE`` etc. in
``mpt._tensor.cosine``) explain the empirical fit. The cells below pin
its decisions on the labelled truth set from that fit, including the
r=2 band where the previous raw-op-count gate over-selected centres by
one to two orders of magnitude in wall time.

Each case is one (K, r, sigma, span_or_period, is_per, expect_centres)
tuple. ``expect_centres`` is what the calibrated model predicts and
also what timing measurements agree with (in the cells verified to
within the crossover neighbourhood).
"""
import math

import numpy as np
import pytest

from mpt._tensor.cosine import (
    _ma_rel_attr_prefers_centres,
    _predicted_centres_wall_ns,
    _predicted_grid_wall_ns,
)


def _make_events(K, span, seed=0):
    """Non-periodic side: pitches uniform on [0, span). Column of Px."""
    rng = np.random.default_rng(seed)
    return np.sort(rng.uniform(0.0, span, size=K)).reshape(K, 1)


# ---------------------------------------------------------------------------
# Non-periodic cells — the r=2 band where the previous gate misfired
# ---------------------------------------------------------------------------

# The old gate said centres for every one of these; the new gate must
# say grid because the closed-form path is 10x-100x slower there.
_NONPER_R2_GRID = [
    # (K, sigma, span_per_side)
    (20, 15.0, 1800.0),
    (30, 15.0, 1800.0),
    (50, 15.0, 1800.0),
    (70, 15.0, 1800.0),
    (30, 30.0, 1800.0),
    (40, 5.0, 1800.0),
    (50, 5.0, 1800.0),
]


@pytest.mark.parametrize("K, sigma, span_per_side", _NONPER_R2_GRID)
def test_nonper_r2_prefers_grid(K, sigma, span_per_side):
    Px = _make_events(K, span_per_side, seed=1)
    Py = _make_events(K, span_per_side, seed=2)
    assert not _ma_rel_attr_prefers_centres(
        Px, Py, sigma, 2, True, False, 0.0
    ), f"K={K} sigma={sigma}: centres path is much slower than grid here"


# Non-periodic cells where centres is genuinely faster.
_NONPER_CENTRES = [
    # (K, r, sigma, span_per_side)
    (5, 2, 15.0, 1800.0),
    (8, 2, 15.0, 1800.0),
    (12, 2, 15.0, 1800.0),
    (6, 3, 15.0, 1800.0),
    (8, 3, 15.0, 1800.0),
    (5, 4, 15.0, 1800.0),
    (6, 4, 15.0, 1800.0),
]


@pytest.mark.parametrize("K, r, sigma, span_per_side", _NONPER_CENTRES)
def test_nonper_prefers_centres(K, r, sigma, span_per_side):
    Px = _make_events(K, span_per_side, seed=1)
    Py = _make_events(K, span_per_side, seed=2)
    assert _ma_rel_attr_prefers_centres(
        Px, Py, sigma, r, True, False, 0.0
    ), f"K={K} r={r} sigma={sigma}: centres path is cheaper here"


# ---------------------------------------------------------------------------
# Periodic cells
# ---------------------------------------------------------------------------

# Periodic cells where the gate must pick grid.
_PER_GRID = [
    # (K, r, sigma, period)
    (30, 2, 15.0, 3600.0),
    (50, 2, 15.0, 3600.0),
    (12, 3, 15.0, 3600.0),
    (10, 3, 5.0, 1200.0),
    (8, 4, 15.0, 3600.0),
]


@pytest.mark.parametrize("K, r, sigma, period", _PER_GRID)
def test_per_prefers_grid(K, r, sigma, period):
    Px = _make_events(K, period, seed=1)
    Py = _make_events(K, period, seed=2)
    assert not _ma_rel_attr_prefers_centres(
        Px, Py, sigma, r, True, True, period
    ), f"K={K} r={r} sigma={sigma} P={period}: grid is cheaper here"


# Periodic cells where centres is cheaper.
_PER_CENTRES = [
    # (K, r, sigma, period)
    (5, 2, 15.0, 3600.0),
    (8, 2, 15.0, 3600.0),
    (6, 3, 15.0, 3600.0),
    (8, 3, 15.0, 3600.0),
    (5, 4, 15.0, 3600.0),
    (6, 4, 15.0, 3600.0),
]


@pytest.mark.parametrize("K, r, sigma, period", _PER_CENTRES)
def test_per_prefers_centres(K, r, sigma, period):
    Px = _make_events(K, period, seed=1)
    Py = _make_events(K, period, seed=2)
    assert _ma_rel_attr_prefers_centres(
        Px, Py, sigma, r, True, True, period
    ), f"K={K} r={r} sigma={sigma} P={period}: centres is cheaper here"


# ---------------------------------------------------------------------------
# Trivial-early-exit cases (invariant under any calibration)
# ---------------------------------------------------------------------------

def test_gate_returns_false_for_absolute_attribute():
    Px = _make_events(20, 1800.0)
    Py = _make_events(20, 1800.0)
    assert _ma_rel_attr_prefers_centres(Px, Py, 15.0, 2, False, False, 0.0) is False


def test_gate_returns_false_for_r_below_two():
    Px = _make_events(20, 1800.0)
    Py = _make_events(20, 1800.0)
    assert _ma_rel_attr_prefers_centres(Px, Py, 15.0, 1, True, False, 0.0) is False


def test_gate_returns_false_when_K_below_r():
    # r=3 needs at least K=3, r=4 at least K=4.
    Px = _make_events(2, 1800.0)
    Py = _make_events(2, 1800.0)
    assert _ma_rel_attr_prefers_centres(Px, Py, 15.0, 3, True, False, 0.0) is False
    Px = _make_events(3, 1800.0)
    Py = _make_events(3, 1800.0)
    assert _ma_rel_attr_prefers_centres(Px, Py, 15.0, 4, True, False, 0.0) is False


def test_gate_returns_false_above_sigma_over_P_threshold():
    # Above σ/P = 0.03 the centres route reads the minimum image while
    # the grid route reads the all-image torus, so the gate must not
    # flip in that band regardless of cost.
    Px = _make_events(8, 100.0)
    Py = _make_events(8, 100.0)
    # sigma/period = 0.10 -> above threshold
    assert not _ma_rel_attr_prefers_centres(
        Px, Py, 10.0, 2, True, True, 100.0
    )


# ---------------------------------------------------------------------------
# Cost model sanity: predictions must be finite and positive.
# ---------------------------------------------------------------------------

@pytest.mark.parametrize("K, r, is_per", [
    (5, 2, False), (15, 2, False), (60, 2, False),
    (6, 3, False), (12, 3, False),
    (5, 4, False), (8, 4, False),
    (5, 2, True), (15, 2, True), (60, 2, True),
    (5, 5, False), (6, 6, False),  # extrapolation regime
])
def test_predicted_walls_finite_and_positive(K, r, is_per):
    c = _predicted_centres_wall_ns(K, r, is_per)
    g = _predicted_grid_wall_ns(K, r, 15.0, 3600.0, is_per)
    assert np.isfinite(c) and c > 0
    assert np.isfinite(g) and g > 0
